File: NetFramework/FlexGrid/VB/HtmlGrid/HtmlEditor.vb
Imports System.IO
Imports System.Text
Imports C1.Win.C1Editor

''' <summary>
''' Class used to edit and render Html in C1FlexGrid cells.
''' </summary>
''' <remarks></remarks>
Public Class HtmlEditor
    Inherits C1Editor

    ' save HTML content when done editing
    Public Function C1EditorGetValue() As Object
        Dim ms As New MemoryStream
        MyBase.SaveXml(ms)
        Dim byteArray As New List(Of Byte)(ms.ToArray)
        byteArray.RemoveRange(0, 3)
        Return Encoding.UTF8.GetString(byteArray.ToArray)
    End Function

    ' assign HTML content when editing starts
    Public Sub C1EditorInitialize(ByVal value As Object, ByVal editorAttributes As IDictionary)
        MyBase.LoadXml(CStr(value), Nothing)
    End Sub

    Public Function C1EditorValueIsValid() As Boolean
        Return MyBase.Enabled
    End Function

    ' support ctrl-B, I, U to implement bold/italic/underline
    Protected Overrides Sub OnKeyDown(ByVal e As KeyEventArgs)
        If ((Not MyBase.SelectedText Is Nothing) AndAlso (e.Modifiers = Keys.Control)) Then
            Select Case e.KeyCode
                Case Keys.B
                    Me.ToggleStyle(FontStyle.Bold)
                    e.Handled = True
                    e.SuppressKeyPress = True
                    Return
                Case Keys.I
                    Me.ToggleStyle(FontStyle.Italic)
                    e.Handled = True
                    e.SuppressKeyPress = True
                    Return
                Case Keys.U
                    Me.ToggleStyle(FontStyle.Underline)
                    e.Handled = True
                    e.SuppressKeyPress = True
                    Return
            End Select
        End If
        MyBase.OnKeyDown(e)
    End Sub

    Private Sub ToggleStyle(ByVal fs As FontStyle)
        Select Case fs
            Case FontStyle.Bold
                If Not MyBase.Selection.IsTagApplied("b") Then
                    MyBase.Selection.ApplyTag("b")
                    Exit Select
                End If
                MyBase.Selection.RemoveTag("b")
                Exit Select
            Case FontStyle.Italic
                If Not MyBase.Selection.IsTagApplied("i") Then
                    MyBase.Selection.ApplyTag("i")
                    Exit Select
                End If
                MyBase.Selection.RemoveTag("i")
                Exit Select
            Case FontStyle.Underline
                If (Not MyBase.Selection.IsStyleApplied("text-decoration") OrElse (String.Compare(MyBase.Selection.GetStyleValue("text-decoration", C1StyleType.Character), "underline", True) <> 0)) Then
                    MyBase.Selection.ApplyStyle("text-decoration", "underline", C1StyleType.Character)
                    Exit Select
                End If
                MyBase.Selection.RemoveStyle("text-decoration", Nothing)
                Exit Select
        End Select
    End Sub

End Class

File: NetFramework/FlexGrid/CS/HtmlGrid/HtmlEditor.cs
using System;
using System.Drawing;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.Windows.Forms;
using System.Runtime.InteropServices;
using C1.Win.C1Editor;
using System.IO;

namespace HtmlGrid
{
	/// <summary>
	/// Class used to edit and render Html in C1FlexGrid cells.
	/// </summary>
	public class HtmlEditor : C1Editor
	{
        //--------------------------------------------------------------------------------
        #region ** ctor

        public HtmlEditor()
        {
        }

        #endregion

        //--------------------------------------------------------------------------------
        #region ** Editing

        public void C1EditorInitialize(object value, IDictionary editorAttributes)
        {
            // assign HTML content when editing starts
            this.LoadXml((string)value, null);
        }

        public object C1EditorGetValue()
        {
            // save HTML content when done editing
            MemoryStream ms = new MemoryStream();
            this.SaveXml(ms);
            List<byte> byteArray = new List<byte>(ms.ToArray());
            byteArray.RemoveRange(0, 3); //remove BOM characters
            return Encoding.UTF8.GetString(byteArray.ToArray());
        }

        public bool C1EditorValueIsValid()
        {
            return Enabled;
        }

        // support ctrl-B, I, U to implement bold/italic/underline
        protected override void OnKeyDown(KeyEventArgs e)
        {
            if (this.SelectedText != null)
            {
                if (e.Modifiers == Keys.Control)
                {
                    switch (e.KeyCode)
                    {
                        case Keys.B:
                            ToggleStyle(FontStyle.Bold);
                            e.Handled = true;
                            e.SuppressKeyPress = true;
                            return;
                        case Keys.I:
                            ToggleStyle(FontStyle.Italic);
                            e.Handled = true;
                            e.SuppressKeyPress = true;
                            return;
                        case Keys.U:
                            ToggleStyle(FontStyle.Underline);
                            e.Handled = true;
                            e.SuppressKeyPress = true;
                            return;
                    }
                }
            }
           base.OnKeyDown(e);
        }
        void ToggleStyle(FontStyle fs)
        {
            switch (fs)
            {
                case FontStyle.Bold:
                    if (Selection.IsTagApplied("b"))
                    {
                        Selection.RemoveTag("b");
                    }
                    else
                    {
                        Selection.ApplyTag("b");
                    }
                    break;
                case FontStyle.Italic:
                    if (Selection.IsTagApplied("i"))
                    {
                        Selection.RemoveTag("i");
                    }
                    else
                    {
                        Selection.ApplyTag("i");
                    }
                    break;
                case FontStyle.Underline:
                    if (Selection.IsStyleApplied("text-decoration") &&
                        string.Compare(Selection.GetStyleValue("text-decoration", C1StyleType.Character), "underline", true) == 0)
                    {
                        Selection.RemoveStyle("text-decoration", null);
                    }
                    else
                    {
                        Selection.ApplyStyle("text-decoration", "underline", C1StyleType.Character);
                    }
                    break;
            }
        }

        #endregion
    }
}
